packages/test/end-to-end-tests/src/test/codePropsal.spec.ts

```diff
@@ -0,0 +1,148 @@
+    };
+
+    let deltaConnectionServer: ILocalDeltaConnectionServer;
+    let opProcessingController: OpProcessingController;
+
+    async function createContainer(factoryEntries: Iterable<[string, IChannelFactory]>): Promise<Container> {
+        const factory = new TestFluidObjectFactory(factoryEntries);
+        const urlResolver = new LocalResolver();
+        const loader: ILoader = createLocalLoader(
+            [[codeDetails, factory],[codeDetails2,factory]],
+             deltaConnectionServer, urlResolver);
+        return createAndAttachContainer(documentId, codeDetails, loader, urlResolver) as any as Container;
+    }
+
+    async function loadContainer(factoryEntries: Iterable<[string, IChannelFactory]>): Promise<Container> {
+        const factory = new TestFluidObjectFactory(factoryEntries);
+        const urlResolver = new LocalResolver();
+        const loader: ILoader = createLocalLoader(
+            [[codeDetails, factory],[codeDetails2,factory]],
+             deltaConnectionServer, urlResolver);
+        return loader.resolve({ url: documentLoadUrl }) as any as Container;
+    }
+
+    let container1: Container;
+    let container2: Container;
+    beforeEach(async () => {
+        deltaConnectionServer = LocalDeltaConnectionServer.create();
+
+        // Create a Container for the first client.
+        container1 = await createContainer([["map", SharedMap.getFactory()]]);
+
+        opProcessingController = new OpProcessingController(deltaConnectionServer);
+        opProcessingController.addDeltaManagers(container1.deltaManager);
+
+        await opProcessingController.process();
+
+        // Load the Container that was created by the first client.
+        container2 = await loadContainer([["map", SharedMap.getFactory()]]);
+        opProcessingController.addDeltaManagers(container1.deltaManager);
+
+        const quorum1 = container1.getQuorum();
+        const quorum2 = container2.getQuorum();
+
+        assert.deepStrictEqual(
+            quorum1.get("code"),
+            codeDetails,
+            "Code proposal in container1 doesn't match");
+
+        assert.deepStrictEqual(
+            quorum2.get("code"),
+            codeDetails,
+            "Code proposal in container2 doesn't match");
+
+        const dataObject1 = await requestFluidObject<ITestFluidObject>(container1, "default");
+        const map1 = await dataObject1.getSharedObject<ISharedMap>("map");
+
+        // BUG BUG quorum.propose doesn't handle readonly, so make sure connection is write
+        while (container1.deltaManager.connectionMode === "read" || !container1.connected) {
+            map1.set("foo","bar");
+            await Promise.all([
+                new Promise((resolve) => container1.connected ? resolve() : container1.once("connect", resolve)),
+                opProcessingController.process(),
+            ]);
+        }
+    });
+
+    it("Code Proposal", async () => {
+        container1.on("contextChanged",(c)=>{
```

Comment:
Is it safe to say that at this point in time the first contextChanged event for `codeDetails1` has already been raised?

Reply:
Or wait, do we no longer raise contextChanged event for initial code proposal? I think that's the case now.

Reply:
we raise it for the for the initial context, but it will already have been raised by this point